File: quide-specification/src/main/java/io/gitlab/arturbosch/quide/platform/QuideDirectory.java
package io.gitlab.arturbosch.quide.platform;

import io.gitlab.arturbosch.quide.validation.Validate;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

/**
 * @author Artur Bosch
 */
public interface QuideDirectory {

	default Path checkDir(Path path) {
		if (Files.notExists(path)) {
			try {
				Files.createDirectories(path);
			} catch (IOException e) {
				throw new UncheckedIOException("Error creating directories for " + path, e);
			}
		}
		return path;
	}

	default Path resolve(String subPath) {
		if (subPath.contains(".")) {
			return checkDir(home()).resolve(subPath);
		} else {
			return checkDir(home().resolve(subPath));
		}
	}

	default Path home() {
		return checkDir(Paths.get(System.getProperty("user.home"), ".quide"));
	}

	default Path pluginsDir() {
		return checkDir(home().resolve("plugins"));
	}

	default Path configurationsDir() {
		return checkDir(home().resolve("configurations"));
	}

	default String getProperty(String key) {
		String property = getProperties().get(key);
		return Validate.isEmpty(property) ? null : property;
	}

	default String getPropertyOrDefault(String key, String defaultValue) {
		String property = getProperties().get(key);
		return Validate.isEmpty(property) ? defaultValue : property;
	}

	Map<String, String> getProperties();

}

File: quide-api/src/main/kotlin/io.gitlab.arturbosch.quide.api/core/QuideDirectory.kt
package io.gitlab.arturbosch.quide.api.core

import java.io.File
import java.io.IOException
import java.util.HashMap

/**
 * @author Artur Bosch
 */
abstract class QuideDirectory(protected val properties: MutableMap<String, String> = HashMap()) : PropertiesAware {

	companion object {
		const val USER_HOME = "user.home"
		const val QUIDE_DIR_NAME = ".quide"
		const val PLUGINS_DIR_NAME = "plugins"
		const val CONFIGURATIONS_DIR_NAME = "configurations"
	}

	val home: File = checkDir(File(System.getProperty(USER_HOME), QUIDE_DIR_NAME))
	val pluginsDir: File = checkDir(resolve(PLUGINS_DIR_NAME))
	val configurationsDir: File = checkDir(resolve(CONFIGURATIONS_DIR_NAME))

	fun resolve(subPath: String): File {
		return checkDir(home.resolve(subPath))
	}

	override fun property(key: String): String? {
		return properties[key]
	}

	override fun propertyOrDefault(key: String, defaultValue: String): String {
		return properties[key] ?: defaultValue
	}

	private fun checkDir(path: File): File {
		if (!path.exists()) {
			if (!path.mkdirs()) {
				throw IOException("Error creating directories for " + path)
			}
		}
		return path
	}

}
